Sources/PromiseKit/Promise.swift

```diff
@@ -766,3 +842,132 @@ extension AnyOptionalType where Self: OptionalType {
   /// See `AnyOptionalType.anyWrappedType`
   public static var anyWrappedType: Any.Type { WrappedType.self }
 }
+
+//==============================================================================
+
+/// Inspired by `SwiftParsec`:
+
+// Operator definitions.
+
+/// Precedence of infix operator for `Promise.chain()`. It has a higher
+/// precedence than the `AssignmentPrecedence` group but a lower precedence than
+/// the `LogicalDisjunctionPrecedence` group.
+precedencegroup FlatMapPrecedence {
+  associativity: left
+  higherThan: AssignmentPrecedence
+  lowerThan: LogicalDisjunctionPrecedence
+}
+
+/// Infix operator for `Promise.chain()`.
+infix operator >>- : FlatMapPrecedence
+
+/// Precedence of infix operator for `Promise.alternative()`. It has a higher
+/// precedence than the `FlatMapPrecedence` group.
+precedencegroup ChoicePrecedence {
+  associativity: left
+  higherThan: FlatMapPrecedence
+}
+
+/// Infix operator for realising choice.
+infix operator <|> : ChoicePrecedence
+
+/// Precedence of infix operators for promise sequencing. It has a higher
+/// precedence than the `ChoicePrecedence` group.
+precedencegroup SequencePrecedence {
+  associativity: left
+  higherThan: ChoicePrecedence
+}
+
+/// Sequence promises, discarding the value of the first promise.
+infix operator *> : SequencePrecedence
+
+/// Sequence promises, discarding the value of the second promise.
+infix operator <* : SequencePrecedence
+
+/// Infix operator for `Promise.apply()`.
+infix operator <*> : SequencePrecedence
+
+/// Infix operator for `Promise.map()`.
+infix operator <^> : SequencePrecedence
+
+
+/// Infix operator for `Promise.chain`.
+///
+/// Often useful when you want to propagate value(s) down the chain.
+///
+/// For example:
+///
+/// f1()
+///  .chain { f2().and($0) }
+///  .chain { f3().and($0) }
+///  .chain { f4($0.1.0) }
+///
+/// can be written as:
+///
+/// f1() >>- { v1 in
+///   f2() >>- { v2 in
+///     f3() >>- { v3 in
+///       f4(v1)
+///     }
+///   }
```

Comment:
I like the clarity of this approach but at the same time, I wonder if we want to do that. Isn't one of the selling points of promises to avoid nesting closures? 🤔

Reply:
I see your point and I agree, in most cases, one should avoid infix notation and use standard chaining. But in those other few cases (when you need a value located far up the chain) I see the value of the proposed approach.
On the other hand, I understand that for someone not familiar with such syntax it may look really strange at the first sight.